### PR TITLE
Rerun of "Add Reagent Grinder/Juicer"

### LINKED ISSUE #2419: Grinder for the cook
grind some wheat, make flour, go brrrrrrrrr

### PR DESCRIPTION
**Pull request by brndd**
Finishing up #2483 after the original contributor dropped it, fixes #2419 

I've cleaned up some stuff but a new review would be handy. Not sure what if anything is missing from this but I don't mind fixing whatever deficiencies there are.

### CODE REVIEW

**Comment by brndd** [comment]
Here's the bounding box I went with, apparently this was a pain point in the original PR.

File: Content.Server/GameObjects/Components/Kitchen/ReagentGrinderComponent.cs [hunk 1]
```diff
@@ -0,0 +1,368 @@
...
+                {
+                    for (int i = chamberCount - 1; i >= 0; i--)
+                    {
+                        var item = chamberContentsArray[i];
+                        if (item.TryGetComponent<SolutionContainerComponent>(out var solution))
+                        {
+                            if (_heldBeaker!.CurrentVolume + solution.CurrentVolume > _heldBeaker!.MaxVolume) continue;
+                            _heldBeaker!.TryAddSolution(solution.Solution);
+                            solution!.RemoveAllSolution();
+                            _chamber.ContainedEntities[i].Delete();
```
Comment (GlassEclipse) [inline Content.Server/GameObjects/Components/Kitchen/ReagentGrinderComponent.cs:334]: I'm not entirely sure this is important but you should probably set _chamber.ContainedEntities[i] to null for cleanness sake (also for the garbage collector?), also do this on line 358 for juicing.
Probably doesn't matter though tbh

File: Content.Server/GameObjects/Components/Kitchen/ReagentGrinderComponent.cs [hunk 2]
```diff
@@ -0,0 +1,368 @@
...
+            {
+                return;
+            }
+            _uiDirty = true;
+            UserInterface?.Toggle(actor.playerSession);
+        }
+
+        public async Task<bool> InteractUsing(InteractUsingEventArgs eventArgs)
+        {
+            if (!Powered)
```
Comment (GlassEclipse) [inline Content.Server/GameObjects/Components/Kitchen/ReagentGrinderComponent.cs:255]: You should be able to put items in the grinder if its unpowered, just not turn it on.

**Comment by metalgearsloth** [comment]
copy+paste last review :)

**Comment by brndd** [comment]
> copy+paste last review :)

I tried going through it but there's lots of steps and the original branch is gone so referencing which parts of it have been fixed and which haven't was kind of a pain. It seemed most of them had been fixed though, gonna have another look later.

**Comment by namespace-Memory** [comment]
how did you do this

**Comment by namespace-Memory** [comment]
such bullshit

**Comment by namespace-Memory** [comment]
based asshole taking my work without permission and taking all the credit for it.

**Comment by brndd** [comment]
The commits are made by you, don't worry.

**Comment by namespace-Memory** [comment]
> The commits are made by you, don't worry.

how did you recover them tho. i want to know so i can avoid this in the future

**Comment by GlassEclipse** [comment]
catastrophic concern

**Comment by namespace-Memory** [comment]
kinda creepy bro ngl

**Comment by RemieRichards** [comment]
Opening a PR is already too late, as the commits in a PR are automatically added to the target repo as soon as the PR is made.

**Comment by namespace-Memory** [comment]
> Opening a PR is already too late, as the commits in a PR are automatically added to the target repo as soon as the PR is made.

they are?

**Comment by namespace-Memory** [comment]
that is such bullshit. you can just steal someones work effortlessly

**Comment by namespace-Memory** [comment]
enjoy the shitcode i guess lol

**Comment by RemieRichards** [comment]
Yes, it's how the "the entire source of youtube-dl is now inside the DMCA repo itself" """trick""" worked.

Also, given that they immediately go into the target repo, they're immediately under that repo's license, so you actually have consented.

**Comment by namespace-Memory** [comment]
> Also, given that they immediately go into the target repo, they're immediately under that repo's license, so you actually have consented.

welp, i wont make that mistake again ty for the knowledge!

**Comment by metalgearsloth** [comment]
> 
> 
> that is such bullshit. you can just steal someones work effortlessly without their consent or knowledge

Could've asked nicely to not have the code used. Also if it were shitcode brndd wouldn't have re-opened it.

**Comment by namespace-Memory** [comment]
>Could've asked nicely
im da evil transphobe who is gonna listen to me. ill try anyway

well @brndd can you please not use my code i would greatly appreciate it if you didnt.

**Comment by brndd** [comment]
If you put it in a PR it's in the tree of the main repo forever afaik. If the commit is from a deleted branch or something there'll be a message saying it's orphaned.

The code isn't that bad and it'd be a shame to let a nearly finished feature go to waste. I assumed you just didn't have the energy to work on it anymore.

> Also, given that they immediately go into the target repo, they're immediately under that repo's license, so you actually have consented.

Even besides that the commit I based this on is explicitly under the MIT license: https://github.com/space-wizards/space-station-14/blob/79da4e0f40e772a0f7497030691949f6a7317699/LICENSE.TXT

**Comment by namespace-Memory** [comment]
>The code isn't that bad and it'd be a shame to let a nearly finished feature go to waste. I assumed you just didn't have the energy to work on it anymore.

wrong

**Comment by namespace-Memory** [comment]
@brndd private message me your paypal and i will pay you not to use my code. i do not want to contribute to ss14 in any shape or form anymore

**Comment by GlassEclipse** [comment]
Cmon memory, you're acting like me whenever I get into ego maniac mode.

**Comment by namespace-Memory** [comment]
> Cmon memory, you're acting like me whenever I get into ego maniac mode.

i was doing just fine before i re entered the cesspool ss14 discord. its my own mistake.

**Comment by Zumorica** [comment]
Let's *not* do this here. Okay? @namespace-Memory if you want to discuss this further, come to the discord.

**Comment by PJB3005** [comment]
Legally, we are under no obligation to stop using your code since you made it public under the terms of the MIT license, which gives us an irrevocable ability to do basically everything with the code as long as it's correctly credited (which it is). (Admittedly there is an "issue" that the MIT license isn't *EXPLICIT* about irrevocability which leads some people to suggest it CAN be revoked but AFAIK this has not been tried in court).

That said, if you don't want your code used I am fine with refusing this PR and requiring it to be rewritten from scratch.

**Comment by namespace-Memory** [comment]
@PJB3005 nah, i settled it with @brndd in DMs. he will continue as he planned.

**Comment by brndd** [comment]
I fixed some edge cases with power checks and made the device usable with no power except for grinding and juicing. Nothing major.

Been testing this a little and it seems to work well enough now. Tried to go through Sloth's reviews in the last PR as best I could and I think they're all addressed now except for this one which I'm gonna need more clarification on:

> If _dirty can be set more than once in a tick then IMO just have an event to add it to the system, then the system runs through any dirty reagentgrinders in a tick when necessary (rather than ticking over every single one). It's not a huge deal given there's probably not a lot of them but it'll be a benefit.

What do you mean by "have an event to add it to the system"? I reckon I could just have the system keep a set of dirty grinders and iterate over that each tick, but I don't see how events would play into this.

**Comment by brndd** [comment]
Alright I did the YAML wrangling, thinking this is about ready now. The juices are missing some features (carrot juice is supposed to heal your eyes for instance) but I dunno if body system does that yet so it'll have to be done in a metabolize pass.

File: Content.Server/GameObjects/Components/Kitchen/ReagentGrinderComponent.cs [hunk 3]
```diff
@@ -0,0 +1,378 @@
...
+        /// of the grinder.
+        /// </summary>
+        private void EjectBeaker(IEntity? user)
+        {
+            if (!HasBeaker || _busy)
+                return;
+
+            _beakerContainer.Remove(_beakerContainer.ContainedEntity);
+
+            if (user == null || !user.TryGetComponent<HandsComponent>(out var hands) || !_heldBeaker!.Owner.TryGetComponent<ItemComponent>(out var item))
```
Comment (metalgearsloth) [inline Content.Server/GameObjects/Components/Kitchen/ReagentGrinderComponent.cs:255]: Remove the null suppressions on _heldBeaker

File: Content.Server/GameObjects/Components/Kitchen/JuiceableComponent.cs [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+    /// </summary>
+    [RegisterComponent]
+    public class JuiceableComponent : Component
+    {
+        public override string Name => "Juiceable";
+        [ViewVariables] public Solution JuiceResultSolution;
+        public override void ExposeData(ObjectSerializer serializer)
+        {
+            base.ExposeData(serializer);
+            serializer.DataField(this, x => JuiceResultSolution, "result", new Solution());
```
Comment (metalgearsloth) [inline Content.Server/GameObjects/Components/Kitchen/JuiceableComponent.cs:19]: should be x.JuiceResultSolution apparently.

**Review by metalgearsloth — APPROVED** [review]
2 minor things

**Review by metalgearsloth — APPROVED** [review]

**Pull request merged**